search-sentiments-map/public/scripts/modal-script.js

```diff
@@ -154,59 +188,81 @@ function highlightWords(
-      snippet = scoreIsPositive ?
-          snippet.replace(new RegExp( '\\b'+ word + '\\b', 'gi'), positiveHighlight) : 
-          snippet.replace(new RegExp('\\b' + word + '\\b', 'gi'), negativeHighlight);
+      snippet = snippet.replace(new RegExp('\\b' + word + '\\b', 'gi'), highlight);
     }
   });
   resultElement.innerHTML += snippet;
 }
 
 /** 
- * Displays the popularity timeline for the current search topic in the 
+ * Sets the popularity timeline for the current search topic in the 
  * selected country's modal.
- * @param {string} countryCode The two-letter code of the selected country.
  */
-function displayPopularityTimeline(countryCode) {
+function setPopularityTimeline() {
   const popularityTimelineElement = document.getElementById('popularity-timeline-tab');
   popularityTimelineElement.innerHTML = '';
-  fetch('/trends/', {
+  postTrendsToGetPopularityTimelineData(getCurrentSearchData().topic, countryCode)
+      .then(timelineJSON => {
+        if (timelineJSON.default.timelineData.length === 0) {
+          popularityTimelineElement.innerText = 
+              'Popularity Timeline is not available for the selected country.';
+        } else {
+          drawPopularityTimeline(
+              timelineJSON.default.timelineData, popularityTimelineElement, 
+              getCurrentSearchData().topic);
+       }
+  });
+}	
+
+/** 
+ * Gets the popularity timeline data from the Google Trends API if it is not
+ * stored in the local cache. Prevents making too many calls to the API if the 
+ * user continuously resizes the window.
+ * @param {string} topic The search topic that the popularity timeline is based on.
+ * @param {string} countryCode Two letter country code for selected country.
+ * @param {Object} popularityTimelineElement The element where the popularity 
+ * timeline is updated.
+ * @return {Promise} The promise which resolves to the the timeline JSON data.
+ */
+function postTrendsToGetPopularityTimelineData(
+    topic, countryCode, popularityTimelineElement) {
+  // Check if country code and topic is in the cache already.
+  if (cachePopularityTimelineData[topic]) {
+    return Promise.resolve(cachePopularityTimelineData[topic][countryCode]);
+  }
+  return fetch('/trends/', { 
     method: 'post',
     headers: {
       'Accept': 'application/json',
       'Content-Type': 'application/json',
     },
     body: JSON.stringify({
-      topic: getCurrentSearchData().topic,
+      topic: topic,
       code: countryCode,
     }),
   })
   .then(interestData => interestData.json())
   .then(timelineJSON => {
-    if (timelineJSON.length === 0) {
-      popularityTimelineElement.innerText = 
-          'Popularity Timeline is not available for the selected country.';
-    } else {
-      drawPopularityTimeline(timelineJSON.default.timelineData, 
-          popularityTimelineElement, getCurrentSearchData().topic);
+    if (!cachePopularityTimelineData[topic]) {
+      cachePopularityTimelineData[topic] = {};
```

Comment:
Why do you need to check this again. Could this have changed since you checked it on line 238?

Reply:
Line 238 (changed) checks if there is something in the cache, but these lines check to see if cachePopularityTimelineData[topic] has been declared previously or not. If it hasn't it has to declare it before setting it to be cachePopularityTimelineData[topic][countryCode] = timelineJSON; in the next lines